File: lib/backend/src/services/ModService.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { spawn } from 'child_process';
import { PROFILES_DIR } from '../utils/paths';
import { downloadFile } from '../utils/file';
import { ProfileService } from './ProfileService';
import { ModManager } from './ModManager';

export class ModService {
    // Helper to get mods from active profile
    static getProfileModsPath(profileId: string) {
        const profileDir = path.join(PROFILES_DIR, profileId);
        const modsDir = path.join(profileDir, 'mods');

        if (!fs.existsSync(modsDir)) {
            fs.mkdirSync(modsDir, { recursive: true });
        }

        return modsDir;
    }

    static generateModId(filename: string) {
        return crypto.createHash('md5').update(filename).digest('hex').substring(0, 8);
    }

    static extractModName(filename: string): string {
        let name = path.parse(filename).name;
        name = name.replace(/-v?\d+\.[\d\.]+.*$/i, '');
        name = name.replace(/-\d+\.[\d\.]+.*$/i, '');
        name = name.replace(/[-_]/g, ' ');
        name = name.replace(/\b\w/g, l => l.toUpperCase());
        return name || 'Unknown Mod';
    }

    static extractVersion(filename: string): string | null {
        const versionMatch = filename.match(/v?(\d+\.[\d\.]+)/);
        return versionMatch ? versionMatch[1] : null;
    }

    static async loadInstalledMods(profileId: string) {
        try {
            // 1. Get current profile config
            const profile = ProfileService.getProfiles()[profileId];
            if (!profile) return [];

            const configMods = profile.mods || [];
            
            const profileModsPath = this.getProfileModsPath(profileId);
            const profileDisabledModsPath = path.join(path.dirname(profileModsPath), 'DisabledMods');

            if (!fs.existsSync(profileModsPath)) fs.mkdirSync(profileModsPath, { recursive: true });
            if (!fs.existsSync(profileDisabledModsPath)) fs.mkdirSync(profileDisabledModsPath, { recursive: true });

            const finalMods: any[] = [];
            const processedFileNames = new Set<string>();

            // 2. Scan disk for current state
            const enabledFiles = fs.readdirSync(profileModsPath).filter(f => f.endsWith('.jar') || f.endsWith('.zip'));
            const disabledFiles = fs.readdirSync(profileDisabledModsPath).filter(f => f.endsWith('.jar') || f.endsWith('.zip'));

            const allFiles = [
                ...enabledFiles.map(f => ({ fileName: f, enabled: true, path: path.join(profileModsPath, f) })),
                ...disabledFiles.map(f => ({ fileName: f, enabled: false, path: path.join(profileDisabledModsPath, f) }))
            ];

            // 3. Process existing config mods
            for (const modConfig of configMods) {
                const fileOnDisk = allFiles.find(f => f.fileName === modConfig.fileName);
                
                if (fileOnDisk) {
                    // Found on disk - update status and path, keep metadata
                    finalMods.push({
                        ...modConfig,
                        enabled: fileOnDisk.enabled,
                        filePath: fileOnDisk.path,
                        missing: false
                    });
                    processedFileNames.add(modConfig.fileName);
                } else {
                    // Not found on disk -> Missing
                    finalMods.push({
                        ...modConfig,
                        filePath: null,
                        missing: true
                    });
                    // We treat missing mods as processed so we don't duplicate them
                    processedFileNames.add(modConfig.fileName);
                }
            }

            // 4. Add new files found on disk that weren't in config
            for (const file of allFiles) {
                if (!processedFileNames.has(file.fileName)) {
                    finalMods.push({
                        id: this.generateModId(file.fileName),
                        fileName: file.fileName,
                        name: this.extractModName(file.fileName),
                        version: this.extractVersion(file.fileName),
                        enabled: file.enabled,
                        filePath: file.path,
                        description: 'Locally installed mod',
                        author: 'Unknown',
                        dateInstalled: new Date().toISOString(),
                        missing: false,
                        manual: true // Flag to indicate manual installation
                    });
                }
            }

            return finalMods;
        } catch (error) {
            console.error('Failed to load mods:', error);
            return [];
        }
    }

    static async toggleMod(profileId: string, fileName: string, enable: boolean) {
        try {
            const profileModsPath = this.getProfileModsPath(profileId);
            const profileDisabledModsPath = path.join(path.dirname(profileModsPath), 'DisabledMods');

            const sourcePath = enable ? path.join(profileDisabledModsPath, fileName) : path.join(profileModsPath, fileName);
            const destPath = enable ? path.join(profileModsPath, fileName) : path.join(profileDisabledModsPath, fileName);

            // 1. Move file
            if (fs.existsSync(sourcePath)) {
                if (!fs.existsSync(path.dirname(destPath))) fs.mkdirSync(path.dirname(destPath), { recursive: true });
                fs.renameSync(sourcePath, destPath);
            } else if (fs.existsSync(destPath)) {
                // Already in target location
            } else {
                return false; // File not found anywhere
            }

            // 2. Update Config
            const profile = ProfileService.getProfiles()[profileId];
            if (profile) {
                const mods = await this.loadInstalledMods(profileId); // Reload to get fresh state including the move
                // Find the mod and force update its enabled state in case loadInstalledMods missed it (unlikely if moved)
                const modIndex = mods.findIndex((m: any) => m.fileName === fileName);
                if (modIndex !== -1) {
                    mods[modIndex].enabled = enable;
                    await ProfileService.updateProfile(profileId, { mods });
                }
            }
            
            // 3. Sync Symlink (Important for game to see changes)
            await ModManager.syncModsForProfile(profileId);

            return true;
        } catch (error) {
            console.error('Toggle mod failed:', error);
            return false;
        }
    }

    static async downloadMod(profileId: string, url: string, fileName: string, modInfo?: any) {
        try {
            const profileModsPath = this.getProfileModsPath(profileId);
            const destPath = path.join(profileModsPath, fileName);

            await downloadFile(url, destPath);

            // Update profile
            const profile = ProfileService.getProfiles()[profileId];
            if (profile) {
                // We manually construct the new mod entry to ensure metadata is captured
                // BEFORE calling loadInstalledMods which might just see a file
                const newMod = {
                    id: modInfo?.id || this.generateModId(fileName),
                    fileName: fileName,
                    name: modInfo?.name || this.extractModName(fileName),
                    version: modInfo?.version || this.extractVersion(fileName),
                    description: modInfo?.description || 'Downloaded Mod',
                    author: modInfo?.author || 'Unknown',
                    curseForgeId: modInfo?.curseForgeId,
                    curseForgeFileId: modInfo?.curseForgeFileId,
                    dateInstalled: new Date().toISOString(),
                    enabled: true,
                    missing: false
                };

                // Get current mods
                const currentMods = profile.mods || [];
                // Remove any existing entry for this filename
                const otherMods = currentMods.filter((m: any) => m.fileName !== fileName);
                
                // Save immediately
                await ProfileService.updateProfile(profileId, { mods: [...otherMods, newMod] });
            }

            // Sync Symlink
            await ModManager.syncModsForProfile(profileId);

            return { success: true, path: destPath };
        } catch (e: any) {
            console.error('Download mod failed:', e);
            return { success: false, error: e.message };
        }
    }

    static async uninstallMod(profileId: string, fileName: string) {
        const profileModsPath = this.getProfileModsPath(profileId);
        const profileDisabledModsPath = path.join(path.dirname(profileModsPath), 'DisabledMods');

        const enabledPath = path.join(profileModsPath, fileName);
        const disabledPath = path.join(profileDisabledModsPath, fileName);

        let deleted = false;
        if (fs.existsSync(enabledPath)) {
            fs.unlinkSync(enabledPath);
            deleted = true;
        } else if (fs.existsSync(disabledPath)) {
            fs.unlinkSync(disabledPath);
            deleted = true;
        }

        if (deleted) {
            // Remove from config
            const profile = ProfileService.getProfiles()[profileId];
            if (profile && profile.mods) {
                const updatedMods = profile.mods.filter((m: any) => m.fileName !== fileName);
                await ProfileService.updateProfile(profileId, { mods: updatedMods });
            }
            
            await ModManager.syncModsForProfile(profileId);
        }

        return deleted;
    }

    // Sync file system state to profile config
    static async syncProfileMods(profileId: string) {
        const mods = await this.loadInstalledMods(profileId);
        await ProfileService.updateProfile(profileId, { mods });

        // Also update the physical symlink to ensure game sees the changes
        await ModManager.syncModsForProfile(profileId);
    }

    static async syncSymlink(profileId: string) {
        return await ModManager.syncModsForProfile(profileId);
    }

    static openModsFolder(profileId: string) {
        const modsDir = this.getProfileModsPath(profileId);
        if (process.platform === 'win32') {
            spawn('explorer', [modsDir]);
        } else if (process.platform === 'darwin') {
            spawn('open', [modsDir]);
        } else {
            spawn('xdg-open', [modsDir]);
        }
        return true;
    }
}

File: lib/backend/src/services/ModService.py
import os
import hashlib
import re
import shutil
from datetime import datetime
from .ModManager import ModManager
from .ProfileService import ProfileService
from .DownloadService import DownloadService

class ModService:
    @staticmethod
    def get_profile_mods_path(profile_id):
        return ModManager.get_profile_mods_path(profile_id)

    @staticmethod
    def generate_mod_id(filename):
        return hashlib.md5(filename.encode()).hexdigest()[:8]

    @staticmethod
    def extract_mod_name(filename):
        name = os.path.splitext(filename)[0]
        # Remove version numbers roughly
        name = re.sub(r'-v?\d+\.[\d\.]+.*$', '', name, flags=re.IGNORECASE)
        name = re.sub(r'-\d+\.[\d\.]+.*$', '', name, flags=re.IGNORECASE)
        name = name.replace('-', ' ').replace('_', ' ')
        return name.title() or 'Unknown Mod'

    @staticmethod
    def extract_version(filename):
        match = re.search(r'v?(\d+\.[\d\.]+)', filename)
        return match.group(1) if match else None

    @classmethod
    def load_installed_mods(cls, profile_id):
        try:
            # 1. Get current profile config
            profile = ProfileService.get_profiles().get(profile_id)
            if not profile:
                return []

            config_mods = profile.get('mods', [])
            
            profile_mods_path = cls.get_profile_mods_path(profile_id)
            profile_disabled_mods_path = os.path.join(os.path.dirname(profile_mods_path), 'DisabledMods')

            if not os.path.exists(profile_mods_path):
                os.makedirs(profile_mods_path, exist_ok=True)
            if not os.path.exists(profile_disabled_mods_path):
                os.makedirs(profile_disabled_mods_path, exist_ok=True)

            final_mods = []
            processed_file_names = set()

            # 2. Scan disk for current state
            def is_mod_file(f):
                return f.endswith('.jar') or f.endswith('.zip')

            enabled_files = [f for f in os.listdir(profile_mods_path) if is_mod_file(f)]
            disabled_files = [f for f in os.listdir(profile_disabled_mods_path) if is_mod_file(f)]

            all_files = []
            for f in enabled_files:
                all_files.append({"fileName": f, "enabled": True, "path": os.path.join(profile_mods_path, f)})
            for f in disabled_files:
                all_files.append({"fileName": f, "enabled": False, "path": os.path.join(profile_disabled_mods_path, f)})

            # 3. Process existing config mods
            for mod_config in config_mods:
                file_name = mod_config.get('fileName')
                file_on_disk = next((f for f in all_files if f['fileName'] == file_name), None)
                
                if file_on_disk:
                    # Found on disk - update status and path, keep metadata
                    mod_entry = mod_config.copy()
                    mod_entry.update({
                        "enabled": file_on_disk['enabled'],
                        "filePath": file_on_disk['path'],
                        "missing": False
                    })
                    final_mods.append(mod_entry)
                    processed_file_names.add(file_name)
                else:
                    # Not found on disk -> Missing
                    mod_entry = mod_config.copy()
                    mod_entry.update({
                        "filePath": None,
                        "missing": True
                    })
                    final_mods.append(mod_entry)
                    processed_file_names.add(file_name)

            # 4. Add new files found on disk that weren't in config
            for file_info in all_files:
                if file_info['fileName'] not in processed_file_names:
                    final_mods.append({
                        "id": cls.generate_mod_id(file_info['fileName']),
                        "fileName": file_info['fileName'],
                        "name": cls.extract_mod_name(file_info['fileName']),
                        "version": cls.extract_version(file_info['fileName']),
                        "enabled": file_info['enabled'],
                        "filePath": file_info['path'],
                        "description": 'Locally installed mod',
                        "author": 'Unknown',
                        "dateInstalled": datetime.utcnow().isoformat() + "Z",
                        "missing": False,
                        "manual": True 
                    })

            return final_mods
        except Exception as error:
            print(f'Failed to load mods: {error}')
            return []

    @classmethod
    def get_mod_details(cls, mod_id):
        from .CurseForgeService import CurseForgeService
        try:
            return CurseForgeService.get_mod(mod_id)
        except Exception as e:
            print(f"Failed to get mod details for {mod_id}: {e}")
            return None

    @classmethod
    def toggle_mod(cls, profile_id, file_name, enable):
        try:
            profile_mods_path = cls.get_profile_mods_path(profile_id)
            profile_disabled_mods_path = os.path.join(os.path.dirname(profile_mods_path), 'DisabledMods')

            source_path = os.path.join(profile_disabled_mods_path if enable else profile_mods_path, file_name)
            dest_path = os.path.join(profile_mods_path if enable else profile_disabled_mods_path, file_name)

            # 1. Move file
            if os.path.exists(source_path):
                dest_dir = os.path.dirname(dest_path)
                if not os.path.exists(dest_dir):
                    os.makedirs(dest_dir, exist_ok=True)
                shutil.move(source_path, dest_path)
            elif os.path.exists(dest_path):
                # Already in target location
                pass
            else:
                return False # File not found anywhere

            # 2. Update Config
            profile = ProfileService.get_profiles().get(profile_id)
            if profile:
                mods = cls.load_installed_mods(profile_id) # Reload to get fresh state
                
                # Find the mod and force update its enabled state
                mod_found = False
                for mod in mods:
                    if mod['fileName'] == file_name:
                        mod['enabled'] = enable
                        mod_found = True
                        break
                
                if mod_found:
                    ProfileService.update_profile(profile_id, {"mods": mods})
            
            # 3. Sync Symlink
            ModManager.sync_mods_for_profile(profile_id)

            return True
        except Exception as error:
            print(f'Toggle mod failed: {error}')
            return False

    @classmethod
    def download_mod(cls, profile_id, url, file_name, mod_info=None):
        try:
            profile_mods_path = cls.get_profile_mods_path(profile_id)
            dest_path = os.path.join(profile_mods_path, file_name)

            DownloadService.download_file(url, dest_path)

            # Update profile
            profile = ProfileService.get_profiles().get(profile_id)
            if profile:
                new_mod = {
                    "id": mod_info.get('id') if mod_info else cls.generate_mod_id(file_name),
                    "fileName": file_name,
                    "name": mod_info.get('name') if mod_info else cls.extract_mod_name(file_name),
                    "version": mod_info.get('version') if mod_info else cls.extract_version(file_name),
                    "description": mod_info.get('description', 'Downloaded Mod'),
                    "author": mod_info.get('author', 'Unknown'),
                    "curseForgeId": mod_info.get('curseForgeId') if mod_info else None,
                    "curseForgeFileId": mod_info.get('curseForgeFileId') if mod_info else None,
                    "dateInstalled": datetime.utcnow().isoformat() + "Z",
                    "enabled": True,
                    "missing": False
                }

                current_mods = profile.get('mods', [])
                # Remove existing entry for this filename
                other_mods = [m for m in current_mods if m.get('fileName') != file_name]
                other_mods.append(new_mod)
                
                ProfileService.update_profile(profile_id, {"mods": other_mods})

            # Sync Symlink
            ModManager.sync_mods_for_profile(profile_id)

            return {"success": True, "path": dest_path}
        except Exception as e:
            print(f'Download mod failed: {e}')
            return {"success": False, "error": str(e)}

    @classmethod
    def uninstall_mod(cls, profile_id, file_name):
        try:
            profile_mods_path = cls.get_profile_mods_path(profile_id)
            profile_disabled_mods_path = os.path.join(os.path.dirname(profile_mods_path), 'DisabledMods')
            
            paths_to_check = [
                os.path.join(profile_mods_path, file_name),
                os.path.join(profile_disabled_mods_path, file_name)
            ]
            
            deleted = False
            for p in paths_to_check:
                if os.path.exists(p):
                    os.remove(p)
                    deleted = True
            
            if deleted:
                # Update config
                profile = ProfileService.get_profiles().get(profile_id)
                if profile:
                    current_mods = profile.get('mods', [])
                    new_mods = [m for m in current_mods if m.get('fileName') != file_name]
                    ProfileService.update_profile(profile_id, {"mods": new_mods})
                
                ModManager.sync_mods_for_profile(profile_id)
                return True
            
            return False
        except Exception as e:
            print(f"Uninstall failed: {e}")
            return False



